doc/compiler/osr/OSR.md

```diff
@@ -20,26 +20,232 @@ OpenJDK Assembly Exception [2].
+The reason why we would like to limit the set of OSR yield points, if possible, is the cost of the bookkeeping associated with each such point as well as the restrictions that they impose on the optimizer.
 
-# potentialOSRPointHelper
-potentialOSRPointHelper is a call that serves as a marker of places where OSR transition is supported. The target bytecode index to be executed in the interpreter can be obtained by adding an offset (an information stored in the call node, the size of OSR point's bytecode in post-execution OSR) to the bytecode index of the helper call node.
+The actual OSR happens at **OSR Induction points**. It can happen at the time of the yield or later, as described in the following sections.
 
-The helper is not a must for OSR, but it can help locating the OSR points that are available at ILGen but disappear due to later transformations. It works as a bookkeeping tool, to track places that are OSR possible.
+Therefore, at a high level, the sequence of events can be described as follows:
 
-# osrFearPointHelper
-This helper marks a place that has been optimized with runtime assumptions and requires protection of OSR mechnism. This means an OSR point is required between the invalidation point and the fear point.
+       VM event => OSR yield => OSR induction => OSR transition
 
-Similar to potentialOSRPointHelper, this helper is not a must for optimizations requiring OSR mechnism. It works as a tool to help locating the actual fear (being part of the program that may be removed or optimized to something else).
+
+# Voluntary and Involuntary OSR
+
+The difference between voluntary vs. involuntary OSR is whether it is induced by the JIT or the VM. In the case of involuntary OSR, the transition is induced by the VM by forcibly transferring control from the VM code to an OSR handler in the compiled code. Such possible transfer of control is modelled as exception control flow in the IL because it behaves similarly. Therefore, every OSR yield point is also an OSR induction point.
+
+Under voluntary OSR, the JIT controls when to trigger an OSR transition by inserting OSR inducing code into the IL.
+
+Each compiled method is compiled in either of the two modes: all possible OSR transitions from that method are voluntary or all OSR transitions are involuntary.
+OMR provides infrastructure for involuntary OSR. Downstream projects can implement voluntary OSR.
+
+# Inducible and Uniducible OSR Yield Points
+
+During voluntary OSR, the JIT can induce OSR after practically every OSR yield point and those points are called **inducible OSR yield points**. However, there are some OSR yield points after which the JIT cannot induce OSR. Those are referred to as **uninducible OSR yield points**. A typical example of an uninducible OSR yield point is when there is [a thunk archetype](https://github.com/eclipse-openj9/openj9/blob/0318ff92766bbbdf36c10a202b08b45ef1b6522f/doc/compiler/README.md?plain=1#L126) present on the inline stack.
```

Comment:
I think the link to OpenJ9 should use the latest version, rather than a link to the README as it existed on a particular day.

Reply:
Oh, you are right. But then it will be a link to the whole document. Is it OK ? @hzongaro

Reply:
Or let me create a MethodHandles.md file and point to it from this document as well as from the README.md

Reply:
That sounds good.  That should make the link more robust and make it obvious that a definition will be needed someday.